### PR TITLE
Add circuit library page

### PR DESCRIPTION
**Pull request by frankharkins**
This PR adds a new page: "Circuit library"

### CODE REVIEW

**Comment by javabster** [comment]
Thanks Frank! I think this still needs a bit of work

For the Circuit Library page we're looking more for a similar structure to the recent blog post (tbh you could even take the blog content and rephrase it to be more documentation and code focused). For this piece I'd expect to see some explanation of the following subsections:
- Standard Gates
- Variational Quantum Circuits
- Data Encoding Circuits
- Time Evolution Circuits
- Arithmetic Circuits

Things like compound gates and parameterised circuits should instead be covered in the new "Circuit Construction" page, issue here: https://github.com/Qiskit/documentation/issues/96

If you would rather turn this PR into one that addresses #96 that's fine by me 😄  or you could split that content out into a second PR and work simultaneously on 96 and 97. I'd also recommend looping in @Cryoris who built a lot of the circuit library and did a lot on the blog

File: docs/build/circuit-library.ipynb [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    "bound_circuit.decompose().draw('mpl')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "78772993-d0da-4651-9211-706e86a59432",
+   "metadata": {},
+   "source": [
+    "- See the [`TwoLocal` documentation](/api/qiskit/qiskit.circuit.library.QFT)\n",
+    "- Read more about parameterized circuits."
```
Comment (frankharkins) [inline docs/build/circuit-library.ipynb:124]: What's the best place to link to about parameterized circuits?

File: docs/build/circuit-library.ipynb
Comment (javabster) [inline docs/build/circuit-library.ipynb:1]: This is looking great! Nice and succinct, good code comments etc.

I cant leave in line comments so heres all feedback in one comment:

> Variational quantum algorithms such as the variational quantum eigensolver (VQE) approximate quantum systems, tweaking gate parameters to find states that have certain properties

Something feels off about this sentence grammatically. It's unclear how this sentence relates to N-local circuits. 

---

> For an example of data encoding circuits in classification, see the [Neural Network Classifier & Regressor tutorial](https://qiskit.org/ecosystem/machine-learning/tutorials/02_neural_network_classifier_and_regressor.html).

I think we should avoid linking to applications modules from the official docs atm as their status is a bit unclear, instead is there a tutorial or course in the learning platform that we could link to instead?

---

For the quantum volume example can we add a link to the arxiv paper explaining how quantum volume is calculated? https://arxiv.org/abs/1811.12926

---

> For a full list of circuits, see the [circuit library API documentation](/api/qiskit/circuit_library).

I think this sentence should be moved to the intro, not left at the end

Comment (frankharkins) [reply docs/build/circuit-library.ipynb:1]: Thank you!

> ... is there a tutorial or course in the learning platform that we could link to instead?

Unfortunately not, but this is a good call; I've added a link to the variational algorithm design course to the paramterized circuits section.

**Review by javabster — APPROVED** [review]
LGTM 🚀

File: docs/build/circuit-library.ipynb [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+    "# Circuit library"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "6f257ff9-15c4-48d8-9503-7f0ab16a91b2",
+   "metadata": {},
+   "source": [
+    "Qiskit includes a library of popular circuits to use as building blocks in your own programs. Using pre-defined circuits saves time researching, writing code, and debugging. The library includes popular circuits in quantum computing, circuits that are difficult to simulate classically, and circuits useful for quantum hardware benchmarking\n",
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:17]:
```suggestion
    "Qiskit includes a library of popular circuits to use as building blocks in your own programs. Using pre-defined circuits saves time researching, writing code, and debugging. The library includes popular circuits in quantum computing, circuits that are difficult to simulate classically, and circuits useful for quantum hardware benchmarking.\n",
```

File: docs/build/circuit-library.ipynb [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "6f257ff9-15c4-48d8-9503-7f0ab16a91b2",
+   "metadata": {},
+   "source": [
+    "Qiskit includes a library of popular circuits to use as building blocks in your own programs. Using pre-defined circuits saves time researching, writing code, and debugging. The library includes popular circuits in quantum computing, circuits that are difficult to simulate classically, and circuits useful for quantum hardware benchmarking.\n",
+    "\n",
+    "This page lists the different categories of circuit the library provides. For a full list of circuits, see the [circuit library API documentation](/api/qiskit/circuit_library)."
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:19]:
```suggestion
    "This page lists the different circuit categories the library provides. For a full list of circuits, see the [circuit library API documentation](/api/qiskit/circuit_library)."
```

File: docs/build/circuit-library.ipynb [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+   "metadata": {},
+   "source": [
+    "## Data encoding circuits\n",
+    "\n",
+    "These parameterized circuits encode data onto quantum states for processing by quantum machine learning algorithms. Some circuits supported by Qiskit are:\n",
+    " - Amplitude encoding, which encodes each number into the amplitude of a basis state. This can store $2^n$ numbers in a single state, but can be costly to implement.\n",
+    " - Basis encoding, which encodes an integer $k$ by preparing the corresponding basis state $|k\\rangle$.\n",
+    " - Angle encoding, which sets each number in the data as a rotation angle in a parameterized circuit.\n",
+    "\n",
+    "The best approach depends upon the specifics of your application. On current quantum computers, however, we often use angle encoding circuits such as the `ZZFeatureMap`."
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb]:
```suggestion
    "## Data-encoding circuits\n",
    "\n",
    "These parameterized circuits encode data onto quantum states for processing by quantum machine learning algorithms. Some circuits supported by Qiskit are:\n",
    " - Amplitude encoding, which encodes each number into the amplitude of a basis state. This can store $2^n$ numbers in a single state, but can be costly to implement.\n",
    " - Basis encoding, which encodes an integer $k$ by preparing the corresponding basis state $|k\\rangle$.\n",
    " - Angle encoding, which sets each number in the data as a rotation angle in a parameterized circuit.\n",
    "\n",
    "The best approach depends upon the specifics of your application. On current quantum computers, however, we often use angle-encoding circuits such as the `ZZFeatureMap`."
```

File: docs/build/circuit-library.ipynb [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "0cf7122d-c3fe-41a6-936c-b3770b33f0f1",
+   "metadata": {},
+   "source": [
+    "## Benchmarking and complexity-theory circuits\n",
+    "\n",
+    "Benchmarking circuits give us a sense of how well our hardware is actually working, and complexity-theory circuits help us understand how difficult the problems we want to solve are.\n",
+    "\n",
+    "For example, the \"quantum volume\" benchmark measures how accurately a quantum computer executes a type of random quantum circuit. The score of the quantum computer increases with the size of the circuit it can reliably run. This takes into account all aspects of the computer, including qubit count, instruction fidelity, qubit connectivity, and the software stack transpiling and post-processing results. Read more about quantum volume in the original [quantum volume paper](https://arxiv.org/abs/1811.12926).\n",
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:243]: Nothing to change now, but QV is changing names probably this week FYI. Since we can't change the name of the original paper, may need to do some wordsmithing here once the name changes.

File: docs/build/circuit-library.ipynb [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    "QuantumVolume(4).decompose().draw('mpl')"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "908e4b4a-5edf-4390-82a0-e755050c2a37",
+   "metadata": {},
+   "source": [
+    "The circuit library also includes circuits believed to be hard to simulate classically, such instantaneous quantum polynomial (IQP) circuits. These circuits sandwich certain diagonal gates (in the computational basis) between blocks of Hadamard gates.\n",
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:277]:
```suggestion
    "The circuit library also includes circuits believed to be hard to simulate classically, such as instantaneous quantum polynomial (IQP) circuits. These circuits sandwich certain diagonal gates (in the computational basis) between blocks of Hadamard gates.\n",
```

File: docs/build/circuit-library.ipynb [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "58b1a7b8-c173-4de8-957a-ca5d58332073",
+   "metadata": {},
+   "source": [
+    "## Arithmetic circuits\n",
+    "\n",
+    "Arithmetic operations are classical functions such as adding integers and bit-wise operations. These may be useful with algorithms such as amplitude estimation for finance applications, and in algorithms like the HHL algorithm, which solves linear systems of equations.\n",
+    "\n",
+    "As an example, let’s try adding two 3-bit numbers using a \"ripple-carry\" circuit to perform in-place addition (`CDKMRippleCarryAdder`). This adder adds two numbers (we'll call them \"A\" and \"B\") and writes the result to the register that held B. In the following example, A=2 and B=3."
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:291]:
```suggestion
    "Arithmetic operations are classical functions, such as adding integers and bit-wise operations. These may be useful with algorithms such as amplitude estimation for finance applications, and in algorithms like the HHL algorithm, which solves linear systems of equations.\n",
    "\n",
    "As an example, let’s try adding two three-bit numbers using a \"ripple-carry\" circuit to perform in-place addition (`CDKMRippleCarryAdder`). This adder adds two numbers (we'll call them \"A\" and \"B\") and writes the result to the register that held B. In the following example, A=2 and B=3."
```

File: docs/build/circuit-library.ipynb [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ee5a64f5-1316-4217-a443-f09cb6d35c19",
+   "metadata": {},
+   "source": [
+    "## Standard gates\n",
+    "\n",
+    "The circuit library also includes standard quantum gates. Some are more fundamental gates (such as the `UGate`),  and others are multi-qubit gates that usually need building from single and two-qubit gates. To add imported gates to your circuit, use the `append` method; the first argument is the gate, and the next argument is a list of qubits to apply the gate to.\n",
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:387]:
```suggestion
    "The circuit library also includes standard quantum gates. Some are more fundamental gates (such as the `UGate`),  and others are multi-qubit gates that usually need building from single- and two-qubit gates. To add imported gates to your circuit, use the `append` method; the first argument is the gate, and the next argument is a list of qubits to apply the gate to.\n",
```

File: docs/build/circuit-library.ipynb [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ee5a64f5-1316-4217-a443-f09cb6d35c19",
+   "metadata": {},
+   "source": [
+    "## Standard gates\n",
+    "\n",
+    "The circuit library also includes standard quantum gates. Some are more fundamental gates (such as the `UGate`),  and others are multi-qubit gates that usually need building from single- and two-qubit gates. To add imported gates to your circuit, use the `append` method; the first argument is the gate, and the next argument is a list of qubits to apply the gate to.\n",
```
Comment (abbycross) [inline docs/build/circuit-library.ipynb:387]:
```suggestion
    "The circuit library also includes standard quantum gates. Some are more fundamental gates (such as the `UGate`), and others are multi-qubit gates that usually need building from single- and two-qubit gates. To add imported gates to your circuit, use the `append` method; the first argument is the gate, and the next argument is a list of qubits to apply the gate to.\n",
```

File: docs/build/_toc.json [hunk 10]
```diff
@@ -8,6 +8,10 @@
     {
       "title": "Build circuits with Qiskit",
       "children": [
+        {
+          "title": "Circuit library",
+          "url": "/build/circuit-library"
+        },
```
Comment (abbycross) [inline docs/build/_toc.json:14]: FYI @frankharkins -- I moved the Circuit library first in its section since that's where @javabster has it listed in the [Build sidebar issue](https://github.com/Qiskit/documentation/issues/100), but feel free to change it back if it was meant to be second in the list!

File: docs/build/circuit-library.ipynb (hunk 5 above)
Comment (frankharkins) [reply docs/build/circuit-library.ipynb:243]: Interesting, what's it changing to?

**Comment by frankharkins** [comment]
Thanks @abbycross!
